File: src/PipelineAtlas.App/src/manifest.ts
// Mirror of the C# manifest contract (PipelineAtlas.Core Model). The viewer reads
// ONLY this — never a live repo or Azure DevOps (CLAUDE.md sec 8.4).

export type NodeType =
  | "entryPipeline"
  | "template"
  | "envConfig"
  | "psModule"
  | "psScript"
  | "test"
  | "doc"
  | "data"
  | "adoEnvironment"
  | "adoResource"
  | "external";

export type NodeStatus = "active" | "legacy-active" | "dormant" | "archived";

export interface Tag {
  kind: "story" | "feature" | "bug";
  id: string;
  url?: string;
}

export interface Flag {
  severity: "secret" | "hardcode" | "techdebt" | "duplication" | "antipattern";
  note: string;
  stepId?: string;
}

export interface Parameter {
  name: string;
  type?: string;
  default?: string;
  description?: string;
}

export interface Node {
  id: string;
  type: NodeType;
  path?: string;
  title: string;
  label?: string; // readable display name (T2b)
  purpose: string;
  clusterId?: string;
  category?: string; // "deployment" | "policy" | ... (T2a)
  trigger?: string;
  pool?: string;
  body?: string; // markdown content, for doc (.md) nodes
  parameters?: Parameter[];
  tags: Tag[];
  flags: Flag[];
  status: NodeStatus;
  source?: "parsed" | "inferred";
}

/** The name to show for a node — its readable label if present, else its title. */
export function displayName(n: Node): string {
  return n.label ?? n.title;
}

export type EdgeKind =
  | "extends"
  | "includesTemplate"
  | "callsScript"
  | "runsOnEnvironment"
  | "deploysTo"
  | "consumesArtifact"
  | "producesArtifact"
  | "gatedBy"
  | "testedBy"
  | "referencesExternal"
  | "documents";

export interface Edge {
  id: string;
  from: string;
  to: string;
  kind: EdgeKind;
  atStepId?: string;
  tags?: Tag[];
}

export type StepKind = "stage" | "job" | "step" | "templateInclude";

export interface Step {
  id: string;
  nodeId: string;
  parentId?: string;
  kind: StepKind;
  name: string;
  doc: string;
  action?: string;
  externalDeps?: string[];
  tags: Tag[];
}

export interface Flow {
  id: string;
  title: string;
  trigger: string;
  path: string[];
  notes?: string;
}

export interface Manifest {
  target: {
    displayName: string;
    description?: string;
    generatedAt: string;
    toolVersion: string;
  };
  nodes: Node[];
  edges: Edge[];
  steps: Step[];
  flows: Flow[];
}

/** Load the manifest served next to the app (dev: public/, prod: the CLI server). */
export async function loadManifest(): Promise<Manifest> {
  const res = await fetch("./manifest.json", { cache: "no-store" });
  if (!res.ok) {
    throw new Error(`Could not load manifest.json (${res.status})`);
  }
  return (await res.json()) as Manifest;
}

// --- small derived helpers ---------------------------------------------------

export const NONE_CLUSTER = "(unclustered)";

export interface ClusterSummary {
  id: string;
  label: string;
  nodeIds: Set<string>;
}

/** Group nodes into clusters (fills a synthetic bucket for unclustered nodes). */
export function clustersOf(m: Manifest): ClusterSummary[] {
  const byId = new Map<string, ClusterSummary>();
  for (const node of m.nodes) {
    const id = node.clusterId ?? NONE_CLUSTER;
    let c = byId.get(id);
    if (!c) {
      c = { id, label: id, nodeIds: new Set() };
      byId.set(id, c);
    }
    c.nodeIds.add(node.id);
  }
  return [...byId.values()].sort((a, b) => a.label.localeCompare(b.label));
}

export function nodeById(m: Manifest): Map<string, Node> {
  return new Map(m.nodes.map((n) => [n.id, n]));
}

/** Distinct tags across the manifest, sorted for the filter control. */
export function allTags(m: Manifest): Tag[] {
  const seen = new Map<string, Tag>();
  for (const n of m.nodes) {
    for (const t of n.tags) seen.set(`${t.kind}:${t.id}`, t);
  }
  return [...seen.values()].sort(
    (a, b) => a.kind.localeCompare(b.kind) || Number(a.id) - Number(b.id),
  );
}

File: src/PipelineAtlas.Core/Model/Manifest.cs
namespace PipelineAtlas.Core.Model;

// The manifest contract — the single shape everything downstream renders from,
// mirrored by Schema/manifest.schema.json (validated on every build). Records are
// immutable; optional (nullable) members are omitted from JSON when null.

public sealed record Tag
{
    public required TagKind Kind { get; init; }
    public required string Id { get; init; }
    public string? Url { get; init; }
}

public sealed record Flag
{
    public required FlagSeverity Severity { get; init; }
    public required string Note { get; init; }
    public string? StepId { get; init; }
}

public sealed record Parameter
{
    public required string Name { get; init; }
    public string? Type { get; init; }
    public string? Default { get; init; }
    public string? Description { get; init; } // seeded from the comment above the parameter
}

public sealed record Node
{
    public required string Id { get; init; }          // stable slug, e.g. "pipeline:DevCI"
    public required NodeType Type { get; init; }
    public string? Path { get; init; }                // target-relative path for file nodes
    public required string Title { get; init; }
    public string? Label { get; init; }               // readable display name (T2b), e.g. "Deploy to DEV (DevCI)"
    public required string Purpose { get; init; }     // plain-language; seeded from header comment
    public string? ClusterId { get; init; }           // from .patlas.json
    public string? Category { get; init; }            // "deployment" | "policy" | ... (T2a)
    public string? Trigger { get; init; }             // "CI: <branch>" | "manual" | "schedule" | "PR gate"
    public string? Pool { get; init; }
    public string? Body { get; init; }                 // markdown content, for doc (.md) nodes (T4b)
    public IReadOnlyList<Parameter>? Parameters { get; init; } // pipeline/template inputs
    public IReadOnlyList<Tag> Tags { get; init; } = [];
    public IReadOnlyList<Flag> Flags { get; init; } = [];
    public required NodeStatus Status { get; init; }
    public NodeSource? Source { get; init; }           // inferred = ADO/missing, not a parsed file
}

public sealed record Edge
{
    public required string Id { get; init; }
    public required string From { get; init; }
    public required string To { get; init; }
    public required EdgeKind Kind { get; init; }
    public string? AtStepId { get; init; }
    public IReadOnlyList<Tag>? Tags { get; init; }
}

public sealed record Step
{
    public required string Id { get; init; }           // e.g. "DevCI/Deploy_Dev/apply-config"
    public required string NodeId { get; init; }
    public string? ParentId { get; init; }             // stage -> job -> step nesting
    public required StepKind Kind { get; init; }
    public required string Name { get; init; }
    public required string Doc { get; init; }          // seeded from inline comments
    public string? Action { get; init; }               // task/command summary
    public IReadOnlyList<string>? ExternalDeps { get; init; }
    public bool? ManualPause { get; init; }             // true = pauses the run for human interaction (ManualValidation/ManualIntervention)
    public IReadOnlyList<Tag> Tags { get; init; } = [];
}

public sealed record Flow
{
    public required string Id { get; init; }
    public required string Title { get; init; }
    public required string Trigger { get; init; }
    public required IReadOnlyList<string> Path { get; init; } // ordered node/step ids traversed
    public string? Notes { get; init; }
}

public sealed record TargetInfo
{
    public required string DisplayName { get; init; }
    public string? Description { get; init; }
    public required string GeneratedAt { get; init; }
    public required string ToolVersion { get; init; }
}

public sealed record Manifest
{
    public required TargetInfo Target { get; init; }
    public required IReadOnlyList<Node> Nodes { get; init; }
    public required IReadOnlyList<Edge> Edges { get; init; }
    public required IReadOnlyList<Step> Steps { get; init; }
    public required IReadOnlyList<Flow> Flows { get; init; }
}
